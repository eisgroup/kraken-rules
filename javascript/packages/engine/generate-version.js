const fs = require('fs')
const path = require('path')
const xmlParser = require('fast-xml-parser')

console.log('----------------------------------');
console.log('generating engine version constant');
console.log('----------------------------------');


const pomPath = path.join(
    __dirname,
    '../',
    '../',
    '../',
    'pom.xml'
)

console.log(`  reading '${pomPath}'`);

const pom = xmlParser.parse(
    fs.readFileSync(
        pomPath,
        { encoding: 'utf8' }
    )
)

console.log(`  version is: '${pom.project.version}'`);

const genFolderPath = path.join(
    __dirname,
    'src',
    'engine',
    'generated'
)

if (!fs.existsSync(genFolderPath)) {
    fs.mkdirSync(genFolderPath)
}

const genPath = path.join(
    __dirname,
    'src',
    'engine',
    'generated',
    'engineVersion.ts'
);

console.log(`  writing to path: '${genPath}'`);

fs.writeFileSync(
    genPath,
    `
// generated code    
// do not change it manually    
export const ENGINE_VERSION = "${pom.project.version}";

`, { encoding: 'utf8' }
)

console.log('  backend version is written');

process.exit(0)