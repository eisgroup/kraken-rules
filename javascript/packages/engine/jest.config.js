module.exports = {
    moduleFileExtensions: ['ts', 'tsx', 'js'],
    transform: {
        '^.+\\.ts$': ['@swc/jest'],
    },
    testMatch: ['**/?(*.)test.ts'],
    verbose: true,
    cacheDirectory: './target/tmp/',
    coverageDirectory: './target/coverage',
    coverageThreshold: {
        global: {
            branches: 93,
            functions: 96,
            lines: 96,
            statements: 96,
        },
    },
    collectCoverageFrom: [
        '**/*[^d\\.].ts',
        '!**/node_modules/**',
        '!**/target/**',
        '!**/__tests__/**',
        '!**/index.ts',
        '!**/**Error**.ts',
        '!**/Reducer.ts', //interface only
        '!**/RulePayloadHandler.ts', //interface only
        '!**/RepositoryClient.ts', //deprecated
        '!**/info/iterators/**', //deprecated
        '!**/config.ts',
        '!**/DevelopmentLogger.ts',
    ],
    testEnvironment: 'node',
    setupFilesAfterEnv: ['./__tests__/console.setup.js', './__tests__/matchers.setup.ts'],
}
