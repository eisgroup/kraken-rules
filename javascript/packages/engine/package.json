{
    "name": "kraken-typescript-engine",
    "version": "1.55.1-stage.0",
    "description": "",
    "main": "target/dist/js/src/index.js",
    "types": "target/dist/definitions/src/index.d.ts",
    "author": "Pavel Surinin <psurinin@eisgroup.com> (http://eisgroup.com)",
    "license": "Apache-2.0",
    "publishConfig": {
        "registry": "https://genesis-npm-release.exigengroup.com/repository/genesis-npm-release/"
    },
    "typedocMain": "src/index.ts",
    "dependencies": {
        "bignumber.js": "9.0.1",
        "declarative-js": "3.8.15",
        "fast-deep-equal": "2.0.1",
        "kraken-engine-api": "1.55.1-stage.0",
        "kraken-model": "1.55.1-stage.0",
        "moment": "2.29.4"
    },
    "devDependencies": {
        "@types/jest": "27.5.1",
        "@types/lodash": "4.14.116",
        "@types/node": "13.9.0",
        "child_process": "1.0.2",
        "fast-xml-parser": "^3.19.0",
        "jest": "27.5.1",
        "jsdom": "11.3.0",
        "jsdom-global": "3.0.2",
        "kraken-jest-matchers": "1.55.1-stage.0",
        "kraken-model-builder": "1.55.1-stage.0",
        "kraken-test-product": "1.55.1-stage.0",
        "kraken-test-product-model-tree": "1.55.1-stage.0",
        "prettier": "1.14.3",
        "rimraf": "3.0.2",
        "ts-node": "4.1.0",
        "typescript": "5.1.6"
    },
    "scripts": {
        "ci-versions": "echo \"typescript version\" && tsc -v && echo \"jest version\" && jest -v",
        "test": "jest -c jest.config.js",
        "test:coverage": "jest -c jest.config.js --coverage",
        "lint": "eslint src __tests__",
        "compile": "node generate-version && tsc --pretty",
        "postcompile": "yarn transpile",
        "transpile": "echo 'no transpile'",
        "del:compiled": "rimraf target",
        "build": "yarn run compile && yarn run lint && yarn run test:coverage",
        "prepare": "yarn run ci-versions && yarn run del:compiled && yarn run compile && yarn run lint -- --fix && yarn run test:coverage",
        "release": "yarn publish"
    },
    "files": [
        "target/dist/definitions",
        "target/dist/js"
    ],
    "gitHead": "efcf1a5e805088ee8af67ef21c2b8ce2fea62eb4"
}