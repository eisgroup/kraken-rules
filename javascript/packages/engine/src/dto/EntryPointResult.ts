/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
