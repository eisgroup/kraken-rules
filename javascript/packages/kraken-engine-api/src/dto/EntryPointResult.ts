/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from './FieldEvaluationResult'
import { RuleEvaluationResults } from './RuleEvaluationResults'
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult

/**
 * Contains evaluation results from specific entry point
 */
export interface EntryPointResult {
    readonly evaluationTimestamp: Date
    readonly ruleTimeZoneId: string

    /**
     * Returns entry point result mapped by field name.
     */
    getFieldResults(): Record<string, FieldEvaluationResult>

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     */
    getAllRuleResults(): RuleEvaluationResult[]

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     * filtered by rule condition evaluation status.
     */
    getApplicableResults(): RuleEvaluationResult[]
}
