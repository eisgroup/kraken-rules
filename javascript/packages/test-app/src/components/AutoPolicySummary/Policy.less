/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

body {
    margin-top: 0px;
}

.input-width {
    width: 90% !important;
    margin-bottom: 15px !important;
    margin-top: 3px !important;
}

.entry-point-name-select {
    margin: 0 10px 0 0;
}

#buttonPanel {
    width: 100%;
    position: fixed;
    top: 0px;
    background: #1890ffe8;
    padding: 16px;
    right: 0px;
    z-index: 1;
    text-align: right;
    box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);
}

#buttonPanel button {
    margin-right: 10px;
}

#buttonPanel div {
    margin-right: 10px;
}

.ContextDefinitionInfo {
    border-radius: 3px;
    padding: 11px;
    background: #90caff59;
    border: 1px solid #2b98fd5e;
    margin: 10px 0px;
}

.sub-entity-box {
    border: 3px solid #80808087;
    padding: 18px;
    margin: 18px;
    border-radius: 3px;
    box-shadow: 0px 0px 20px -3px rgba(0, 0, 0, 0.5);
    background: url('https://www.toptal.com/designers/subtlepatterns/patterns/dot-grid.png');
}
.sub-entity-box > h2 {
    color: rgb(45, 154, 255);
    text-align: center;
    text-transform: capitalize;
}
