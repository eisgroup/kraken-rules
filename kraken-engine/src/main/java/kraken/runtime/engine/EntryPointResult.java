package kraken.runtime.engine;

import kraken.annotations.API;
import kraken.runtime.engine.dto.FieldEvaluationResult;
import kraken.runtime.engine.dto.RuleEvaluationResult;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Contains evaluation results from specific entry point
 *
 * @author rimas
 * @since 1.0
 */
@API
public class EntryPointResult {

    /**
     * Field evaluation results, mapped to field id string
     */
    private Map<String, FieldEvaluationResult> fieldResults;

    private transient LocalDateTime evaluationTimeStamp;
    /**
     * Default constructor
     */
    public EntryPointResult() {
        fieldResults = new HashMap<>();
        evaluationTimeStamp = LocalDateTime.now();
    }

    public EntryPointResult(Map<String, FieldEvaluationResult> fieldResults, LocalDateTime evaluationTimeStamp) {
        this.fieldResults = new HashMap<>(fieldResults);
        this.evaluationTimeStamp = evaluationTimeStamp;
    }

    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     *
     * @return  list of all contained {@link RuleEvaluationResult}s
     */
    public List<RuleEvaluationResult> getAllRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .collect(Collectors.toList());
    }

    public List<RuleEvaluationResult> getApplicableRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .filter(r -> r.getConditionEvaluationResult().isApplicable())
                .collect(Collectors.toList());
    }

    public LocalDateTime getEvaluationTimeStamp() {
        return evaluationTimeStamp;
    }

    public Map<String, FieldEvaluationResult> getFieldResults() {
        return fieldResults;
    }

}
