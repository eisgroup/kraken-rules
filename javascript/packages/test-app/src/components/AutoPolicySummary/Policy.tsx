/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import * as React from "react";

import "antd/lib/radio/style";
import "antd/lib/input/style";
import "antd/lib/input-number/style";
import "antd/lib/date-picker/style";
import "antd/lib/button/style";
import "antd/lib/spin/style";
import "antd/lib/select/style";
import "antd/lib/collapse/style";
import "antd/lib/row/style";
import "antd/lib/list/style";
import "antd/lib/notification/style";
import "antd/lib/slider/style";

import { Button, Spin, Row, Select, Collapse, Popover, Radio } from "antd";

import { SingleField, ManyFields } from "../core/field/SingleField";
import { renderers } from "../core/RenderInputFunctions";
import { entryPointNames } from "../../common/constants";

import { AutoPolicyViewState, EvaluationResponse } from "./Policy.container";
import { TransactionDetails } from "./TransactionDetails";
import { PolicyDetails } from "./PolicyDetails";
import { TermDetails } from "./TermDetails";
import { AccessTrackInfo } from "./AccessTrackInfo";
import { CreditCardInfo } from "./CreditCardInfo";
import { Insured } from "./Insured";
import { Vehicle } from "./Vehicle";
import { Party } from "./Party";
import { UIResultModal } from "./modals/UIResultModal";
import { BackendResultModal } from "./modals/BackendResultModal";
import { ModelModal } from "./modals/ModelModal";
import { ImportModal } from "./modals/ImportModal";

import { TestProduct } from "kraken-test-product";
import domain = TestProduct.kraken.testproduct.domain;
import { ErrorBoundary } from "../../ErrorBoundary";
import { ContextDefinitionInfo } from "../core/field/ContextDefinitionInfo";
import { EntityBox } from "../core/EntityBox";
import { Referer } from "./Referer";
import { ExpressionTreeModal } from "./modals/ExpressionTreeModal";

export interface AutoPolicyCallbackFunctionProps {
    onClickShowExpressionModal: () => void;
    onClickShowUIResultModal: () => void;
    onClickShowBackendResultModal: () => void;
    onClickShowModelModal: () => void;
    onClickShowImportModal: () => void;
    onImportableModelSubmit: () => void;
    onImportableModelChange: (e: React.FormEvent<HTMLTextAreaElement>) => void;
    onValidationEntryPointChange: (validationEntryPointName: string) => void;
    validate: () => void;
    onClickExportCurrentState: () => void;
    onClickResetUI: () => void;
    onStateChange: (state: string) => void;
    onPolicyChange: (e: React.FormEvent<HTMLInputElement>) => void;
    onCreatedFromPolicyRevChange: (e: React.ReactText) => void;
    onPoliciesChange: (e: string[]) => void;
    onTransDetailsChange: (det: domain.TransactionDetails) => void;
    onPolicyDetailsChange: (det: domain.PolicyDetail) => void;
    onTermDetailsChange: (det: domain.TermDetails) => void;
    onAccessTrackInfoChange: (accessTrackInfo: domain.AccessTrackInfo) => void;
    onCreditCardInfoChange: (creditCardInfo: domain.CreditCardInfo) => void;
    onInsuredChange: (insured: domain.Insured) => void;
    onVehicleChange: (changedVehicleIndex: Number) => (det: domain.Vehicle) => void;
    onPartyChange: (party: domain.Party) => void;
    onRefToCustomerChange: (e: React.FormEvent<HTMLInputElement>) => void;
    onClickResetEntryPoint: () => void;
    onRefererChange: (r: domain.Referer) => void;
}

export interface AutoPolicyProps extends AutoPolicyCallbackFunctionProps {
    autoPolicyViewState: AutoPolicyViewState & EvaluationResponse;
}

export class Policy extends React.Component<AutoPolicyProps> {
    render(): JSX.Element {
        const { autoPolicyViewState } = this.props;
        return (
            <Spin spinning={false} size="large">
                {/* <Spin spinning={autoPolicyViewState.isLoading} size="large"> */}
                <div>
                    <b>{autoPolicyViewState.buildInfo && autoPolicyViewState.buildInfo.date}</b>
                    <UIResultModal
                        onClickShowModal={this.props.onClickShowUIResultModal}
                        modalVisibility={autoPolicyViewState.isUIResultModalVisible}
                        fieldResults={autoPolicyViewState.fieldResults}
                        allRuleResults={autoPolicyViewState.allRuleResults}
                        validationStatus={autoPolicyViewState.validationStatus}
                    />
                    <BackendResultModal
                        onClickShowModal={this.props.onClickShowBackendResultModal}
                        modalVisibility={autoPolicyViewState.isBackendResultModalVisible}
                        backendRawResult={autoPolicyViewState.backendRawResult}
                        backendDefaultResult={autoPolicyViewState.backendDefaultResult}
                        backendValidationResult={autoPolicyViewState.backendValidationResult}
                    />
                    <ModelModal
                        onClickShowModal={this.props.onClickShowModelModal}
                        modalVisibility={autoPolicyViewState.isModelModalVisible}
                        modelBeforeEvaluation={autoPolicyViewState.modelBeforeEvaluation}
                        modelAfterEvaluation={autoPolicyViewState.modelAfterEvaluation}
                        modelCurrentState={autoPolicyViewState.modelCurrentState}
                    />
                    <ExpressionTreeModal
                        modalVisibility={autoPolicyViewState.isExpressionModalVisible}
                        onClickShowModal={this.props.onClickShowExpressionModal}
                    />
                    <ImportModal
                        modalVisibility={autoPolicyViewState.isImportModalVisible}
                        onClickShowModal={this.props.onClickShowImportModal}
                        onImportableModelSubmit={this.props.onImportableModelSubmit}
                        onImportableModelChange={this.props.onImportableModelChange}
                        importableModelState={autoPolicyViewState.importableModelState}
                    />
                    <div id="buttonPanel">
                        <Select
                            className="entry-point-name-select"
                            defaultValue={entryPointNames.UI}
                            onChange={this.props.onValidationEntryPointChange}
                            style={{ width: 200 }}
                        >
                            {Object.keys(entryPointNames).map((entryPointName: string) => (
                                <Select.Option key={entryPointName} value={entryPointName}>
                                    {entryPointName}
                                </Select.Option>
                            ))}
                        </Select>
                        <Button
                            loading={this.props.autoPolicyViewState.isLoading}
                            type="dashed"
                            icon="code-o"
                            onClick={this.props.validate}
                        >
                            Validate
                        </Button>
                        <Button type="dashed" icon="eye-o" onClick={this.props.onClickShowUIResultModal}>
                            UI result
                        </Button>
                        <Button
                            type={this.props.autoPolicyViewState.isBackendFailed ? "danger" : "dashed"}
                            icon="eye-o"
                            onClick={this.props.onClickShowBackendResultModal}
                        >
                            Backend result
                        </Button>
                        <Button type="dashed" icon="eye-o" onClick={this.props.onClickShowModelModal}>
                            Model
                        </Button>
                        <Button type="dashed" icon="experiment" onClick={this.props.onClickShowExpressionModal}>
                            Expression
                        </Button>
                        <Button type="dashed" icon="upload" onClick={this.props.onClickShowImportModal}>
                            Import model
                        </Button>
                        <Button type="dashed" icon="download" onClick={this.props.onClickExportCurrentState}>
                            Export state
                        </Button>
                        <Button type="danger" icon="delete" onClick={this.props.onClickResetUI}>
                            Reset form
                        </Button>
                        <Button type="danger" icon="reload" onClick={this.props.onClickResetEntryPoint}>
                            Reset EntryPoint
                        </Button>
                        <br />
                        <span>
                            Status <b id="status">{autoPolicyViewState.isLoading ? "Loading" : "OK"}</b>&nbsp;
                        </span>
                        <br />
                    </div>
                    <ErrorBoundary>
                        <EntityBox title="Policy" >
                            <ContextDefinitionInfo contextName="Policy" />
                            <Row>
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.state}
                                    contextName="Policy"
                                    modelFieldName="state"
                                    onChange={this.props.onStateChange}
                                    // tslint:disable-next-line: max-line-length
                                    info="dimension: state, use 'Dimensional' entrypoint to get rules based on that dimension"
                                    renderInput={renderers.select}
                                    selections={["CA", "AZ"]}

                                />
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.policyNumber}
                                    contextName="Policy"
                                    modelFieldName="policyNumber"
                                    onChange={this.props.onPolicyChange}
                                    renderInput={renderers.input}
                                />
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.createdFromPolicyRev}
                                    contextName="Policy"
                                    modelFieldName="createdFromPolicyRev"
                                    onChange={this.props.onCreatedFromPolicyRevChange}
                                    renderInput={renderers.inputNumber}
                                />
                                <SingleField
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.refToCustomer}
                                    contextName="Policy"
                                    modelFieldName="refToCustomer"
                                    onChange={this.props.onRefToCustomerChange}
                                    renderInput={renderers.inputDisabled}
                                    info="This field is readonly"
                                />
                                <ManyFields
                                    id={autoPolicyViewState.autoPolicy.id}
                                    value={autoPolicyViewState.autoPolicy.policies}
                                    contextName="Policy"
                                    modelFieldName="policies"
                                    onChange={this.props.onPoliciesChange}
                                    renderInput={renderers.input}
                                />
                            </Row>
                            <TransactionDetails
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.transactionDetails}
                                onChange={this.props.onTransDetailsChange}
                            />
                            <PolicyDetails
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.policyDetail}
                                onChange={this.props.onPolicyDetailsChange}
                            />
                            <TermDetails
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.termDetails}
                                onChange={this.props.onTermDetailsChange}
                            />
                            <AccessTrackInfo
                                id={autoPolicyViewState.autoPolicy.id}
                                value={autoPolicyViewState.autoPolicy.accessTrackInfo}
                                onChange={this.props.onAccessTrackInfoChange}
                            />
                        </EntityBox>
                        <EntityBox title="Referer" >
                            <ContextDefinitionInfo contextName="Referer" />
                            <Referer
                                id={autoPolicyViewState.autoPolicy.referer.id}
                                value={autoPolicyViewState.autoPolicy.referer}
                                onChange={this.props.onRefererChange}
                            />
                        </EntityBox>
                        <EntityBox title="CreditCardInfo" >
                            <ContextDefinitionInfo contextName="CreditCardInfo" />
                            <CreditCardInfo
                                id={autoPolicyViewState.autoPolicy.billingInfo.creditCardInfo.id}
                                value={autoPolicyViewState.autoPolicy.billingInfo.creditCardInfo}
                                onChange={this.props.onCreditCardInfoChange}
                            />
                        </EntityBox>
                        <EntityBox title="Insured" >
                            <ContextDefinitionInfo contextName="Insured" />
                            <Insured
                                id={autoPolicyViewState.autoPolicy.insured.id}
                                value={autoPolicyViewState.autoPolicy.insured}
                                onChange={this.props.onInsuredChange}
                            />
                        </EntityBox>
                        <EntityBox title="Risk Items" >
                            <Collapse>
                                {autoPolicyViewState.autoPolicy.riskItems.map((riskItem, index) => (
                                    <Collapse.Panel header={"Vehicle " + index} key={riskItem.id}>
                                        <Vehicle
                                            id={autoPolicyViewState.autoPolicy.riskItems[index].id}
                                            value={autoPolicyViewState.autoPolicy.riskItems[index]}
                                            onChange={this.props.onVehicleChange(index)}
                                        />
                                    </Collapse.Panel>
                                ))}
                            </Collapse>
                        </EntityBox>
                        <EntityBox title="Party" >
                            <Party
                                id={autoPolicyViewState.autoPolicy.parties[0].id}
                                value={autoPolicyViewState.autoPolicy.parties[0]}
                                onChange={this.props.onPartyChange}
                            />
                        </EntityBox>
                    </ErrorBoundary>
                </div>
            </Spin >
        );
    }
}
