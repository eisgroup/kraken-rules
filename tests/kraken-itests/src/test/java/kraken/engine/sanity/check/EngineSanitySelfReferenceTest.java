/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package kraken.engine.sanity.check;

import kraken.runtime.engine.EntryPointResult;
import kraken.testproduct.domain.Policy;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static io.github.jsonSnapshot.SnapshotMatcher.validateSnapshots;
import static kraken.testing.matchers.KrakenMatchers.*;
import static org.junit.Assert.assertThat;
import static kraken.test.KrakenItestMatchers.matchesSnapshot;

/**
 * @author psurinin@eisgroup.com
 * @since 1.0.30
 */
public class EngineSanitySelfReferenceTest extends SanityEngineBaseTest {

    @BeforeClass
    public static void beforeAll() {
        start();
    }

    @AfterClass
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    public void shouldReferenceSameContextAsRuleDefined() {
        Policy policy = new Policy();
        policy.setPolicyNumber("P1");
        final EntryPointResult result = engine.evaluate(policy, "SelfReference");

        assertThat(result, hasRuleResults(3));
        assertThat(result, hasNoValidationFailures());
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, matchesSnapshot());
    }
}
