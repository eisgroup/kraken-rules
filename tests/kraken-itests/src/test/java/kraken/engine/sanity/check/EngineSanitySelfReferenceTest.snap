kraken.engine.sanity.check.EngineSanitySelfReferenceTest.shouldReferenceSameContextAsRuleDefined=[
  {
    "Policy:-1:policyNumber": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "policyNumber",
        "fieldPath": "policyNumber"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "SelfImplicit",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "SelfCCR",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "SelfThis",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]